CMake: Build fails

```
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:35:25: error: no template named 'variant' in namespace 'std'
using Parameters = std::variant<GPGParameters, AESParameters>;
                   ~~~~~^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:54:75: error: use of undeclared identifier 'Parameters'
        processPlainText(std::vector<uint8_t> plain_text, std::shared_ptr<Parameters>&& params) = 0;
                                                                          ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:65:77: error: use of undeclared identifier 'Parameters'
        processCipherText(std::vector<uint8_t> cipher_text, std::shared_ptr<Parameters>&& params) = 0;
                                                                            ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:75:71: error: use of undeclared identifier 'Parameters'
    static std::shared_ptr<Cipher> get(Scheme scheme, std::shared_ptr<Parameters>&& init_params={});
                                                                      ^
In file included from /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/gpgcrypto.cpp:30:
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/gpgcrypto.h:48:75: error: use of undeclared identifier 'Parameters'
        processPlainText(std::vector<uint8_t> plain_text, std::shared_ptr<Parameters>&& params) override;
                                                                          ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/gpgcrypto.h:51:77: error: use of undeclared identifier 'Parameters'
        processCipherText(std::vector<uint8_t> cipher_text, std::shared_ptr<Parameters>&& params) override;
                                                                            ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/gpgcrypto.cpp:67:93: error: use of undeclared identifier 'Parameters'
std::vector<uint8_t> GPG::processPlainText(std::vector<uint8_t> plain_text, std::shared_ptr<Parameters>&& params)
                                                                                            ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/gpgcrypto.cpp:88:74: error: use of undeclared identifier 'Parameters'
GPG::processCipherText(std::vector<uint8_t> cipher_text, std::shared_ptr<Parameters>&& params)
                                                                         ^
8 errors generated.
[2/4] /usr/bin/c++  -DPACKAGE_NAME=\"dpaste\" -DVERSION=\"0.3.3\" -I/usr/local/include -I/usr/local/include/curlpp -I/usr/local/include/glibmm-2.4 -I/usr/local/lib/glibmm-2.4/include -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include/sigc++-2.0 -I/usr/local/lib/sigc++-2.0/include -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -I/usr/local/include/nlohmann -DBUFFERSIZE=1024 -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -I/usr/local/include/nlohmann -DBUFFERSIZE=1024   -std=gnu++14 -MD -MT CMakeFiles/dpaste.dir/src/bin.cpp.o -MF CMakeFiles/dpaste.dir/src/bin.cpp.o.d -o CMakeFiles/dpaste.dir/src/bin.cpp.o -c /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.cpp
FAILED: CMakeFiles/dpaste.dir/src/bin.cpp.o 
/usr/bin/c++  -DPACKAGE_NAME=\"dpaste\" -DVERSION=\"0.3.3\" -I/usr/local/include -I/usr/local/include/curlpp -I/usr/local/include/glibmm-2.4 -I/usr/local/lib/glibmm-2.4/include -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include/sigc++-2.0 -I/usr/local/lib/sigc++-2.0/include -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -I/usr/local/include/nlohmann -DBUFFERSIZE=1024 -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -I/usr/local/include/nlohmann -DBUFFERSIZE=1024   -std=gnu++14 -MD -MT CMakeFiles/dpaste.dir/src/bin.cpp.o -MF CMakeFiles/dpaste.dir/src/bin.cpp.o.d -o CMakeFiles/dpaste.dir/src/bin.cpp.o -c /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.cpp
In file included from /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.cpp:29:
In file included from /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:34:
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:35:25: error: no template named 'variant' in namespace 'std'
using Parameters = std::variant<GPGParameters, AESParameters>;
                   ~~~~~^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:54:75: error: use of undeclared identifier 'Parameters'
        processPlainText(std::vector<uint8_t> plain_text, std::shared_ptr<Parameters>&& params) = 0;
                                                                          ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:65:77: error: use of undeclared identifier 'Parameters'
        processCipherText(std::vector<uint8_t> cipher_text, std::shared_ptr<Parameters>&& params) = 0;
                                                                            ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/cipher.h:75:71: error: use of undeclared identifier 'Parameters'
    static std::shared_ptr<Cipher> get(Scheme scheme, std::shared_ptr<Parameters>&& init_params={});
                                                                      ^
In file included from /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.cpp:29:
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:70:76: error: no member named 'Parameters' in namespace 'dpaste::crypto'
    std::string paste(std::vector<uint8_t>&& data, std::unique_ptr<crypto::Parameters>&& params);
                                                                   ~~~~~~~~^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:71:81: error: no member named 'Parameters' in namespace 'dpaste::crypto'
    std::string paste(std::stringstream&& input_stream, std::unique_ptr<crypto::Parameters>&& params) {
                                                                        ~~~~~~~~^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:73:54: error: no member named 'Parameters' in namespace 'dpaste::crypto'
                std::forward<std::unique_ptr<crypto::Parameters>>(params));
                                             ~~~~~~~~^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:73:75: error: expected a type
                std::forward<std::unique_ptr<crypto::Parameters>>(params));
                                                                          ^
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:74:6: error: expected '}'
    }
     ^
```

Revision ```0.3.3-55-gf27df83```.
FreeBSD 11.2 amd64

@sim590 It uses ```-std=gnu++14```, not ```-std=c++17``` that is needed to enable ```c++17``` features.
Additionally, the ```config.h``` header is missing:
```
/usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:31:10: fatal error: 'config.h' file not found
#include "config.h"
         ^~~~~~~~~~
1 error generated.
```

So it is currently broken.

> @sim590 It uses `-std=gnu++14`, not `-std=c++17` that is needed to enable `c++17` features.

The check [is missing in CMake][cmake], but [is not missing in Autotools][autotools]. Therefore, you're using CMake to build. Please, include those information when you're supplying a bug. I have to say that I may have ignored CMake for a bit when doing my last changes. That is wrong and I'll fix it.

[cmake]: https://github.com/sim590/dpaste/blob/master/CMakeLists.txt#L7
[autotools]: https://github.com/sim590/dpaste/blob/master/configure.ac#L16

> Additionally, the `config.h` header is missing:
>
> ```
> /usr/ports/misc/dpaste/work/dpaste-0.3.3-55-gf27df83/src/bin.h:31:10: fatal error: 'config.h' file not found
> #include "config.h"
>          ^~~~~~~~~~
> 1 error generated.
> ```
> So it is currently broken.

I see that it's missing a

```c++
#ifdef HAVE_CONFIG_H
#endif
```

Again, another CMake related issue. I'm on it.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,6 +32,8 @@ list(APPEND dpaste_HEADERS
     src/bin.h
     src/gpgcrypto.h
     src/log.h
+	src/cipher.h
+	src/aescrypto.h
 )
 list(APPEND dpaste_SOURCES
     src/main.cpp
@@ -41,6 +43,8 @@ list(APPEND dpaste_SOURCES
     src/bin.cpp
     src/gpgcrypto.cpp
     src/log.cpp
+	src/cipher.cpp
+	src/aescrypto.cpp
 )
 
 #################################

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,7 +4,7 @@ set(dpaste_VERSION 0.3.3)
 add_definitions(-DVERSION="${dpaste_VERSION}")
 add_definitions(-DPACKAGE_NAME="dpaste")
 
-set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD 17)
 set(THREADS_PREFER_PTHREAD_FLAG ON)
 
 ###################

--- a/src/bin.h
+++ b/src/bin.h
@@ -28,7 +28,9 @@
 #include <map>
 #include <utility>
 
+#ifdef HAVE_CONFIG_H
 #include "config.h"
+#endif
 #include "node.h"
 #include "http_client.h"
 #include "cipher.h"

--- a/src/log.cpp
+++ b/src/log.cpp
@@ -23,7 +23,9 @@
 #include <stdio.h>
 #include <stdarg.h>
 
+#ifdef HAVE_CONFIG_H
 #include "config.h"
+#endif
 #include "log.h"
 
 static constexpr const char* DPASTE_MSG_PREFIX = "DPASTE: ";

--- a/src/node.h
+++ b/src/node.h
@@ -30,7 +30,9 @@
 #include <opendht/rng.h>
 #include <opendht/callbacks.h>
 
+#ifdef HAVE_CONFG_H
 #include "config.h"
+#endif
 
 namespace dpaste {
 #ifdef DPASTE_TEST

